pkg/leech/pipeline.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// Prepare connects the internal pipeline implementation to the beam
+// scope object associated with the primary beam Pipeline.
+func (pipe *Pipeline) Prepare(scope beam.Scope) {
+	// BigQuery table notation is not consistent so it needs represented in a few different
+	// formats to appease the BigQuery client.
+	eventsTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.EventsProjectID, pipe.cfg.EventsTable)
+	leechTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	leechTableColonNotation := fmt.Sprintf("%s:%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	batchSize := pipe.cfg.BatchSize
+
+	var event EventRecord
+	query := fmt.Sprintf(sourceQuery, eventsTableDotNotation, leechTableDotNotation, batchSize, batchSize)
+	// step 1: query BigQuery for unprocessed events
+	col := bigqueryio.Query(scope, pipe.cfg.LeechProjectID, query, reflect.TypeOf(event), bigqueryio.UseStandardSQL())
+	// step 2: process the events in parallel, ingesting logs
+	res := beam.ParDo(scope, &ingestLogsFn{pipe: pipe}, col)
+	// step 3: write all of the results back to BigQuery
+	bigqueryio.Write(scope, pipe.cfg.LeechProjectID, leechTableColonNotation, res)
+}
+
+// readPrivateKey reads a PEM encouded private key from a string.
+func readPrivateKey(privateKeyContent string) (*rsa.PrivateKey, error) {
+	parsedKey, _, err := jwk.DecodePEM([]byte(privateKeyContent))
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode PEM formated key:  %w", err)
+	}
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert to *rsa.PrivateKey (got %T)", parsedKey)
+	}
+	return privateKey, nil
+}
+
+// ProcessElement is a DoFn implementation that reads workflow logs from GitHub
+// and stores them in Cloud Storage.
+func (f *ingestLogsFn) ProcessElement(event EventRecord) LeechRecord {
+	ctx := context.Background()
+	logger := logging.FromContext(ctx)
+
+	gcsPath := fmt.Sprintf("gs://%s/%s/%s/artifacts.tar.gz", f.pipe.cfg.LogsBucketName, event.RepositorySlug, event.DeliveryID)
+	result := LeechRecord{
+		DeliveryID:       event.DeliveryID,
+		ProcessedAt:      time.Now(),
+		WorkflowURI:      event.WorkflowURL,
+		GitHubActor:      event.GitHubActor,
+		OrganizationName: event.OrganizationName,
+		RepositoryName:   event.RepositoryName,
+		RepositorySlug:   event.RepositorySlug,
+		LogsURI:          gcsPath,
+		Status:           "SUCCESS",
+	}
+	logger.Infof("processing element", "DeliveryID", event.DeliveryID, "event", event, "result", result)
+	if err := f.handleMessage(ctx, event.RepositoryName, event.LogsURL, gcsPath); err != nil {
+		// Expired logs can never be retrieved, mark them as gone and move on
+		if errors.Is(err, errLogsExpired) {
+			logger.Infof("logs for workflow not available", "DeliveryID", event.DeliveryID)
+			result.Status = "NOT_FOUND"
+		} else {
+			// Other failures can be retried
+			// @TODO(bradegler): These are currently filtered out of the query, need to figure
```

Comment:
Can we make this a bug?

Reply:
I can create an issue, sure, I was planning on working on this tomorrow though was mainly just capturing notes for myself.